File: src/aq.c
/* aq.c — one-producer/one-consumer access-unit FIFO with byte-cap backpressure. */
#include <pthread.h>
#include <stdlib.h>
#include <string.h>
#include "aq.h"

void aq_init(au_queue *q) {
    memset(q, 0, sizeof *q);
    pthread_mutex_init(&q->m, NULL);
    pthread_cond_init(&q->not_full, NULL);
    pthread_cond_init(&q->not_empty, NULL);
}

/* Destroy the sync objects (call once the producer/consumer are done). Pairs with
 * aq_init so a re-init on the next playback isn't UB-on-an-initialized-mutex. */
void aq_destroy(au_queue *q) {
    pthread_mutex_destroy(&q->m);
    pthread_cond_destroy(&q->not_full);
    pthread_cond_destroy(&q->not_empty);
}

/* Producer: append one AU (copies `len` bytes). Blocks while the queue is over
 * AQ_MAX_BYTES unless aborting. Returns 0 on success, -1 if aborting or OOM. */
int aq_push(au_queue *q, const unsigned char *data, int len,
            long long pts, int key, int es) {
    au_node *n = (au_node *)malloc(sizeof(au_node) + (size_t)len);
    if (!n) return -1;
    n->next = NULL; n->pts = pts; n->len = len; n->key = key; n->es = es;
    memcpy(n->data, data, (size_t)len);

    pthread_mutex_lock(&q->m);
    while (q->queued_bytes > AQ_MAX_BYTES && !q->abort)
        pthread_cond_wait(&q->not_full, &q->m);
    if (q->abort) { pthread_mutex_unlock(&q->m); free(n); return -1; }
    if (q->tail) q->tail->next = n; else q->head = n;
    q->tail = n;
    q->queued_bytes += len;
    pthread_cond_signal(&q->not_empty);
    pthread_mutex_unlock(&q->m);
    return 0;
}

/* Consumer: pop the next AU (caller frees it), or NULL if empty. Never blocks.
 * *eof_out is set to 1 when the producer has finished and the queue is drained. */
au_node *aq_pop(au_queue *q, int *eof_out) {
    pthread_mutex_lock(&q->m);
    au_node *n = q->head;
    if (n) {
        q->head = n->next;
        if (!q->head) q->tail = NULL;
        q->queued_bytes -= n->len;
        pthread_cond_signal(&q->not_full);
    }
    if (eof_out) *eof_out = (!q->head && q->eof);
    pthread_mutex_unlock(&q->m);
    return n;
}

void aq_set_eof(au_queue *q) {
    pthread_mutex_lock(&q->m);
    q->eof = 1;
    pthread_cond_signal(&q->not_empty);
    pthread_mutex_unlock(&q->m);
}

/* Ask the producer to stop blocking and bail (teardown). */
void aq_abort(au_queue *q) {
    pthread_mutex_lock(&q->m);
    q->abort = 1;
    pthread_cond_broadcast(&q->not_full);
    pthread_cond_broadcast(&q->not_empty);
    pthread_mutex_unlock(&q->m);
}

long aq_bytes(au_queue *q) {
    pthread_mutex_lock(&q->m);
    long b = q->queued_bytes;
    pthread_mutex_unlock(&q->m);
    return b;
}

File: rust-modules/src/aq.rs
//! One-producer / one-consumer access-unit FIFO with byte-cap backpressure (was
//! src/aq.c). The player consumer allocates an `AuQueue` and passes `&q`; `aq_pop`
//! hands back a malloc'd `AuNode` the caller reads and frees. AuNode uses a flexible
//! array member (`data[]`); AuQueue embeds pthread objects — both repr(C) so the
//! layouts match glibc.
use std::os::raw::{c_int, c_long, c_uchar, c_void};
use std::ptr;

#[repr(C)]
pub struct AuNode {
    next: *mut AuNode,
    pts: i64,
    len: c_int,
    key: c_int,
    es: c_int,
    // flexible array `data[]` follows immediately after `es` (see node_data)
}

#[repr(C)]
pub struct AuQueue {
    head: *mut AuNode,
    tail: *mut AuNode,
    queued_bytes: c_long,
    max_bytes: c_long, // per-queue backpressure cap; the caller chooses it (see aq_new)
    eof: c_int,
    abort: c_int,
    m: libc::pthread_mutex_t,
    not_full: libc::pthread_cond_t,
    not_empty: libc::pthread_cond_t,
}

/// pointer to the au_node's flexible `data[]` (offset = right after `es`)
#[inline]
unsafe fn node_data(n: *mut AuNode) -> *mut u8 {
    (n as *mut u8).add(core::mem::offset_of!(AuNode, es) + core::mem::size_of::<c_int>())
}

/// crate-internal: has the consumer asked the producer to stop? (the demuxer checks this)
#[inline]
pub(crate) unsafe fn aq_is_aborted(q: *const AuQueue) -> bool {
    !q.is_null() && (*q).abort != 0
}

/// crate-internal: read a popped node's fields (es, key, pts, len, data ptr) —
/// the player engine's feed loop needs them without re-exposing the raw offsets.
#[inline]
pub(crate) unsafe fn au_fields(n: *mut AuNode) -> (c_int, c_int, i64, c_int, *const u8) {
    ((*n).es, (*n).key, (*n).pts, (*n).len, node_data(n) as *const u8)
}

/// allocate + initialize a boxed queue with an explicit backpressure byte cap (mirrors
/// `stream::http_stream_boxed`; the two-lane feed gives each ES its own cap).
pub(crate) fn aq_new(cap: c_long) -> Box<AuQueue> {
    let mut q: Box<AuQueue> = Box::new(unsafe { std::mem::zeroed() });
    aq_init_cap(&mut *q, cap);
    q
}

/// initialize an already-allocated queue with an explicit backpressure byte cap.
fn aq_init_cap(q: *mut AuQueue, cap: c_long) {
    if q.is_null() { return; }
    unsafe {
        ptr::write_bytes(q as *mut u8, 0, core::mem::size_of::<AuQueue>());
        (*q).max_bytes = cap;
        libc::pthread_mutex_init(ptr::addr_of_mut!((*q).m), ptr::null());
        libc::pthread_cond_init(ptr::addr_of_mut!((*q).not_full), ptr::null());
        libc::pthread_cond_init(ptr::addr_of_mut!((*q).not_empty), ptr::null());
    }
}

pub(crate) fn aq_destroy(q: *mut AuQueue) {
    if q.is_null() { return; }
    unsafe {
        libc::pthread_mutex_destroy(ptr::addr_of_mut!((*q).m));
        libc::pthread_cond_destroy(ptr::addr_of_mut!((*q).not_full));
        libc::pthread_cond_destroy(ptr::addr_of_mut!((*q).not_empty));
    }
}

/// Producer: append one AU (copies `len` bytes). Blocks over the queue's `max_bytes` cap
/// unless aborting. Returns 0 on success, -1 if aborting or OOM.
pub(crate) fn aq_push(q: *mut AuQueue, data: *const c_uchar, len: c_int,
                          pts: i64, key: c_int, es: c_int) -> c_int {
    if q.is_null() || len < 0 { return -1; }
    unsafe {
        let n = libc::malloc(core::mem::size_of::<AuNode>() + len as usize) as *mut AuNode;
        if n.is_null() { return -1; }
        (*n).next = ptr::null_mut();
        (*n).pts = pts;
        (*n).len = len;
        (*n).key = key;
        (*n).es = es;
        if len > 0 && !data.is_null() {
            ptr::copy_nonoverlapping(data, node_data(n), len as usize);
        }
        libc::pthread_mutex_lock(ptr::addr_of_mut!((*q).m));
        while (*q).queued_bytes > (*q).max_bytes && (*q).abort == 0 {
            libc::pthread_cond_wait(ptr::addr_of_mut!((*q).not_full), ptr::addr_of_mut!((*q).m));
        }
        if (*q).abort != 0 {
            libc::pthread_mutex_unlock(ptr::addr_of_mut!((*q).m));
            libc::free(n as *mut c_void);
            return -1;
        }
        if !(*q).tail.is_null() {
            (*(*q).tail).next = n;
        } else {
            (*q).head = n;
        }
        (*q).tail = n;
        (*q).queued_bytes += len as c_long;
        libc::pthread_cond_signal(ptr::addr_of_mut!((*q).not_empty));
        libc::pthread_mutex_unlock(ptr::addr_of_mut!((*q).m));
        0
    }
}

/// Consumer: pop the next AU (caller frees it) or NULL. Never blocks.
/// *eof_out = 1 when the producer finished and the queue is drained.
pub(crate) fn aq_pop(q: *mut AuQueue, eof_out: *mut c_int) -> *mut AuNode {
    if q.is_null() { return ptr::null_mut(); }
    unsafe {
        libc::pthread_mutex_lock(ptr::addr_of_mut!((*q).m));
        let n = (*q).head;
        if !n.is_null() {
            (*q).head = (*n).next;
            if (*q).head.is_null() {
                (*q).tail = ptr::null_mut();
            }
            (*q).queued_bytes -= (*n).len as c_long;
            libc::pthread_cond_signal(ptr::addr_of_mut!((*q).not_full));
        }
        if !eof_out.is_null() {
            *eof_out = if (*q).head.is_null() && (*q).eof != 0 { 1 } else { 0 };
        }
        libc::pthread_mutex_unlock(ptr::addr_of_mut!((*q).m));
        n
    }
}

pub(crate) fn aq_set_eof(q: *mut AuQueue) {
    if q.is_null() { return; }
    unsafe {
        libc::pthread_mutex_lock(ptr::addr_of_mut!((*q).m));
        (*q).eof = 1;
        libc::pthread_cond_signal(ptr::addr_of_mut!((*q).not_empty));
        libc::pthread_mutex_unlock(ptr::addr_of_mut!((*q).m));
    }
}

pub(crate) fn aq_abort(q: *mut AuQueue) {
    if q.is_null() { return; }
    unsafe {
        libc::pthread_mutex_lock(ptr::addr_of_mut!((*q).m));
        (*q).abort = 1;
        libc::pthread_cond_broadcast(ptr::addr_of_mut!((*q).not_full));
        libc::pthread_cond_broadcast(ptr::addr_of_mut!((*q).not_empty));
        libc::pthread_mutex_unlock(ptr::addr_of_mut!((*q).m));
    }
}

pub(crate) fn aq_bytes(q: *mut AuQueue) -> c_long {
    if q.is_null() { return 0; }
    unsafe {
        libc::pthread_mutex_lock(ptr::addr_of_mut!((*q).m));
        let b = (*q).queued_bytes;
        libc::pthread_mutex_unlock(ptr::addr_of_mut!((*q).m));
        b
    }
}
